### PR TITLE
Forgot password

### PR DESCRIPTION
**Pull request by mafuvadze**
## Description
I added an API endpoint to respond to user password reset requests. A reset token is sent to the user's on file email address via email from no-reply@gthc.io

## Motivation and Context
If someone forgets their password, they need a way to be reset it and access their account.

## How Has This Been Tested?

- Ask Anesu or Aman for the password to the email (or ask Anesu to send you the secrets.yml file)
- run 'yarn dev:start'
- in an API client like postman or ARC, make a POST request to 'http://localhost:5000/api/v1/user/forgot_password' with the body {user_email: ""}
- Check your email
- See email with password reset link (link to doesn't do anything yet)

## Screenshots (if appropriate):
screen shot 2019-01-02 at 6 41 32 pm

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by amanmibra** [comment]
I will review this later tonight

File: app/javascript/src/actions/user.js [hunk 1]
```diff
@@ -38,6 +38,18 @@ const checkSession = () => (
   })
 );
 
+const initiatePasswordReset = () => {
```
Comment (amanmibra) [inline app/javascript/src/actions/user.js]: You should complete integration into redux or break into another PR

File: config/secrets.yml [hunk 2]
```diff
@@ -19,9 +19,13 @@
 
 development:
   secret_key_base: 5fc0bc43d7cfb12126b2c9567f35d96d0b384eeb94f3014fb20b12214fb839db11166e7341e2c0acbf650c55d0944b4d05dbf8f47263204397fee769768d27b8
+  gthc_email: 'no-reply@gthc.io'
+  gthc_email_password: 'wDSg7J5Ak4ZJajkK'
```
Comment (amanmibra) [inline config/secrets.yml]: The password should not be public. Maybe let's use ENV for now until we complete our Rails 5.2 update onto `staging`.

File: config/environments/production.rb [hunk 3]
```diff
@@ -94,4 +94,19 @@
...
   # Serve assets under public
   # config.serve_static_assets = true
+
+  config.action_mailer.default_url_options = { :host => 'gthc.io' }
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.perform_deliveries = true
+  config.action_mailer.raise_delivery_errors = true
+
+  ActionMailer::Base.smtp_settings = {
+    :user_name => Rails.application.secrets[:gthc_email],
```
Comment (amanmibra) [inline config/environments/production.rb]: Like I mentioned in my other comment, let's use environment variables, let's stay consistent with other non-public values, and we can integrate confidentials across the board once we go 5.2.

File: app/javascript/src/actions/user.js (hunk 1 above)
Comment (amanmibra) [reply app/javascript/src/actions/user.js]: Or, you can acknowledge this with a TODO like you did in `forgot_password`.

**Review by amanmibra — COMMENTED** [review]
Overall looks good, but take a look at my suggestions before pushing. Decide what you want to do and then go ahead merge.

File: config/secrets.yml (hunk 2 above)
Comment (mafuvadze) [reply config/secrets.yml]: Oh you're right I forgot to update this
good catch

File: config/environments/production.rb (hunk 3 above)
Comment (mafuvadze) [reply config/environments/production.rb]: Ok I'll hold of this PR until we have 5.2 on staging

Comment (amanmibra) [reply config/environments/production.rb]: No build it with ENV for now, and we can have a whole separate branch for moving to credentials if we sit it fit our needs better. Parts of this PR is also necessary on admin as it also uses Devise and has a forgot password option.

Comment (amanmibra) [reply config/environments/production.rb]: @mafuvadze

Comment (Vinitowinit) [reply config/environments/production.rb]: @mafuvadze Any update my friend??

File: config/secrets.yml (hunk 2 above)
Comment (amanmibra) [reply config/secrets.yml]: Fix this

File: config/environments/production.rb (hunk 3 above)
Comment (amanmibra) [reply config/environments/production.rb]: Can you please update this?

Comment (mafuvadze) [reply config/environments/production.rb]: @mafuvadze updated

File: app/controllers/api/v1/users_controller.rb [hunk 4]
```diff
@@ -106,6 +106,7 @@ def forgot_password
     validate_forgot_password_params
     @user = User.find_by_email(params[:user_email]);
     if @user
+      puts 'yippy'
```
Comment (amanmibra) [inline app/controllers/api/v1/users_controller.rb]: this needs to be removed

File: app/javascript/src/actions/utils/crud.js [hunk 5]
```diff
@@ -2,6 +2,7 @@ import axios from 'axios';
 import { push } from 'react-router-redux';
 
 const crud = request => (dispatch) => {
+  console.log(request.url);
```
Comment (amanmibra) [inline app/javascript/src/actions/utils/crud.js]: needs to be removed

**Review by amanmibra — COMMENTED** [review]
Left some logs and puts, I will go ahead and remove them myself. If all works fine locally with me, then I will merge it myself.

**Review by amanmibra — CHANGES_REQUESTED** [review]
Sorry, but this is far from finished. Your PR is causing issues to pre-existing components. Some issues I noticed from just playing around for 5 minutes.

1. Redux actions not even exported... [FIXED] (come on man how are you gonna try to merge this to staging like this)
2. There needs to be error handling for unknown/incorrect emails on `/forgot_password`. 
3. When changing passwords, please stop it from logging in automatically after finishing the form because there are empty redux states that are usually filled in by the normal `POST /login` call. You can see the issue and try it, login normally and check the `user` redux state and see how it has a user's user, team, and captain data in the `user` redux state but when you are logged in through the forgot password. It does not fill those in, so if you try to go pages like `app/team`, it will break.
4. Design of web pages (lowest priority but also, we have standards, we can't just hand black pages. I can understand the rails view, but the `/forgot_password` should be improved.

**Comment by amanmibra** [comment]
@mafuvadze 
Since the Semantic UI is loaded already with react, which is actually just a portal of its html components, you can still access sematic ui html. Therefore, you can design the rails view using sematic ui except it would be in HTML. Use these docs to help you:

https://semantic-ui.com/

**Comment by amanmibra** [comment]
In my most recent commits, I added some semantic ui classes to show that it can be done, and hopefully give you a head start in the right direction.

File: config/secrets.yml (hunk 2 above)
Comment (amanmibra) [inline config/secrets.yml]: bro @mafuvadze

**Review by amanmibra — COMMENTED** [review]
You forgot to remove the `secrets.yml`

**Comment by amanmibra** [comment]
Bro how many times do I have to say redirect the app to the `login` page once they change their password, DO NOT SIGN THEM IN @mafuvadze

**Comment by amanmibra** [comment]
Please test this more thoroughly, there are still a lot of problems
